test/unit/core/main.js

```diff
@@ -31,6 +31,91 @@ suite('Core', function() {
         p5.prototype._registeredMethods.init = originalInit;
       }
     });
+    test('should register and call before and after "preload" hooks', function() {
+      return new Promise(resolve => {
+        let beforePreloadCalled = false;
+        let preloadCalled = false;
+        let afterPreloadCalled = false;
+
+        p5.prototype.registerMethod('beforePreload', () => {
+          beforePreloadCalled = true;
+        });
+
+        p5.prototype.registerMethod('preload', () => {
+          assert.equal(beforePreloadCalled, true);
+          preloadCalled = true;
+        });
+
+        p5.prototype.registerMethod('afterPreload', () => {
+          if(beforePreloadCalled && preloadCalled)
+            afterPreloadCalled = true;
+        });
+
+        myp5 = new p5(function(sketch) {
+          sketch.preload = () => {};
+          sketch.setup = () => {
+            assert.equal(afterPreloadCalled, true);
+          };
+          resolve();
+        });
+      });
+    });
+    test('should register and call before and after "setup" hooks', function() {
+      return new Promise(resolve => {
+        let beforeSetupCalled = false;
+        let setupCalled = false;
+        let afterSetupCalled = false;
+
+        p5.prototype.registerMethod('beforeSetup', () => {
+          beforeSetupCalled = true;
+        });
+
+        p5.prototype.registerMethod('setup', () => {
+          assert.equal(beforeSetupCalled, true);
+          setupCalled = true;
+        });
+
+        p5.prototype.registerMethod('afterSetup', () => {
+          if(beforeSetupCalled && setupCalled)
+            afterSetupCalled = true;
+        });
+
+        myp5 = new p5(function(sketch) {
+          sketch.setup = () => {};
+          sketch.draw = () => {
+            assert.equal(afterSetupCalled, true);
+            resolve();
+          };
+        });
+      });
+    });
+    test('should register and call pre and post "draw" hooks', function() {
+      return new Promise(resolve => {
+        let preDrawCalled = false;
+        let drawCalled = false;
+        let postDrawCalled = false;
+
+        p5.prototype.registerMethod('pre', () => {
+          preDrawCalled = true;
+        });
+
+        p5.prototype.registerMethod('draw', () => {
+          assert.equal(preDrawCalled, true);
+          drawCalled = true;
+        });
+
+        p5.prototype.registerMethod('post', () => {
+          if(preDrawCalled && drawCalled)
+            postDrawCalled = true;
+        });
+
+        myp5 = new p5(function(sketch) {
+          sketch.draw = () => {};
+        });
+        assert.equal(postDrawCalled, true);
```

Comment:
Getting a `postDraw` assertion is going to be a bit tough. The reason why we can't do assertions outside of `setup` and `draw` here is that those actually get run *first*, before `draw` is called: when we define `draw`, we're only defining it, not calling it, and it gets called asynchronously later after `preload` and `setup` finish. So if we `resolve` outside of it, then likely we exit before `draw` gets called at all.

As an alternative: the first time `draw` gets called, `sketch.frameCount` will be 1. So we could do something like this, to make sure it gets run after the first draw call:
```js
sketch.draw = () => {
  if (sketch.frameCount === 2) {
    assert.equal(postDrawCalled, true);
    resolve();
  }
};
```